src/thunderbots/software/ai/hl/stp/evaluation/enemy_threat.cpp

```diff
@@ -157,3 +159,125 @@ std::optional<std::pair<int, std::optional<Robot>>> Evaluation::getNumPassesToRo
     // state. Therefore, we return an std::nullopt
     return std::nullopt;
 }
+
+std::vector<Evaluation::EnemyThreat> Evaluation::sortThreatsInDecreasingOrder(
+    std::vector<Evaluation::EnemyThreat> threats)
+{
+    // A lambda function that implements the '<' operator for the EnemyThreat struct
+    // so it can be sorted. Lower threats are "less than" higher threats.
+    auto enemyThreatLessThanComparator = [](const EnemyThreat &a, const EnemyThreat &b) {
+        // Robots with the ball are more threatening than robots without the ball, and
+        // robots with the ball are the most threatening since they can shoot or move
+        // the ball towards our net
+        if (a.has_ball && !b.has_ball)
+        {
+            return false;
+        }
+        else if (!a.has_ball && b.has_ball)
+        {
+            return true;
+        }
+        // If both robots have the ball, the robot with a worse shot on our net is less
+        // threatening (although this case is unlikely to happen since usually only 1
+        // robot can have the ball at a time)
+        else if (a.has_ball && b.has_ball)
+        {
+            return a.best_shot_angle < b.best_shot_angle;
+        }
+        else
+        {
+            // If neither robot has the ball, the robot that takes longer to reach via
+            // passing is less threatening
+            if (a.num_passes_to_get_possession < b.num_passes_to_get_possession)
+            {
+                return false;
+            }
+            else if (a.num_passes_to_get_possession > b.num_passes_to_get_possession)
+            {
+                return true;
+            }
+            else
+            {
+                // Finally, if both robots can be reached in the same number of passes,
+                // the robot with a smaller view of the net is considered less
+                // threatening. The reason we use goal_angle here rather than the
+                // best_shot_angle is that the goal_angle doesn't change if the robot is
+                // blocked from shooting (eg. by a defender). This makes the evaluation
+                // more stable since the value won't change drastically as our robots
+                // move into defensive positions and change the best_shot_angle. If we had
+                // fewer robots than the enemy team and were using the best_shot_angle,
+                // defenders could oscillate between enemies since when the defender
+                // blocks one enemy, the unblocked one becomes more threatening and the
+                // defender would then move there.
+                return a.goal_angle < b.goal_angle;
```

Comment:
I see why you chose goal_angle here, but I feel like this would cause too many robots to block the enemy. Adding defenders to this enemy robot doesn't really reduce the threat in this case so friendly robots may just keep getting assigned to the same bot. (Unless we only assign 1 robot per threat)

Reply:
We should only assign 1 robot per threat.

The point of using `goal_angle` **is** so that adding defenders doesn't change the threat. See my comment for a more detailed explanation, but basically it could cause oscillation in defenders.

I think we'll get a better idea of if this is completely right once I write some defensive tactics that use this evaluation. If things don't work as expected we can re-evaluate then, so I'll probably leave this for now.